### ISSUE TITLE
EasyBlock to create a Conda package as part of a Python bundle?

### ISSUE THREAD

**Issue by kelseymh** [easybuilders/easybuild-easyblocks#2016]
I recently created local EasyConfigs to build several Python packages needed for our group -- numpy, rootpy and root_numpy.  One of the working environments we may have to use is Anaconda.  After loading the modules I had created, I tried using `conda install --use-local` to get them into an Anaconda environment.  It failed, and complained that it couldn't find the module I wanted in "other channels" (specifically on repo.continuum.io), which I didn't want searched anyway.

Apparently Conda has its own package building system, and it looks for objects built that way when you do |conda install|.  Does it make sense to have an EasyBlock that can take care of that for us, maybe along similar lines to PythonPackage?

It seems like the information needed for Conda's "meta.yaml" file is all present in the EasyConfig (name, version, source info (Git or PyPI), etc.), and the build scripts (bld.bat and build.sh) are hardcoded.

**Comment by ocaisa** [comment]
Hi @kelseymh , my python-fu is not as good as other maintainers but my understanding is that our use of PYTHONPATH would mean that our Python modules should be picked up inside conda already (which some people would see as a negative thing) unless `PYTHONPATH` is actively unset. When you initialise your conda env this may well be the case, but if you load a module after initialisation that module would set PYTHONPATH and that would then be picked up inside conda. 

There is already a conda easyblock that can install conda packages (see https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/q/QIIME2/QIIME2-2019.7.eb and https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/r/Roary/Roary-3.12.0.eb for examples) but these are used on top of the `system` toolchain since you will fill your environment with conda stuff in an uncontrolled way (from an EasyBuild perspective) once you start using this. I don't think it would be possible to do combinations at a higher level, since we wouldn't have sufficient control over what conda is doing.

So, as far as I can tell, the "right" thing to do is to load your required conda environment first built in the standard way (or using what EB supports), then load all the modules that you require from EasyBuild. The EasyBuild stuff will then take precedence over the conda stuff so you get the missing things you needed from conda but the optimised software from EB.

**Comment by kelseymh** [comment]
What it sounds like you're suggesting is that if we have to load a Conda environment (virtualenv), we should make it as bare minimum as possible (in our case, just "notebook" for Jupyter).  Once we're in, we can use Python `import XYZ` and it will find our EasyBuild stuff (via `$PYTHONPATH`).  

That's an acceptable solution, and much better than populating the EasyBuild modules with a bunch of Conda builds.  Thank you!

**Comment by boegel** [comment]
One side note here: usually when you load a module for a (bundle of) Python package(s) installed with EasyBuild, you'll also get a `Python` module loaded that was built with EasyBuild.

With `conda`, you also get a new `python` command.

With that in mind, I doubt that trying to puzzle together stuff provided by `conda` with modules installed with EasyBuild is going to work out well, unless you're really careful (like making sure the same major/minor Python version is used on both sides).

Long story short, it may cause more problems than it fixes...

**Comment by kelseymh** [comment]
@boegel Yes, I agree with your conclusion, "it may cause more problems than it fixes..."  I closed this issue.  It was extremely helpful to discuss it, and to realize that the Conda system really doesn't play well with mixed Python versions.

For our particular use case, the Texas A&M HPRC folks are supporting a plain Python module for their Jupyter Notebook interface, along with the ability of users to specify an additional module (such as my group's Bundle) to load.  With that combination, we can now do simple "import ROOT" or "import numpy" directives in Jupyter and get exactly the version we want from our Bundle.

This avoids the Conda version mixing issues, since the plain Python module can be superseded safely by our own Bundle or its contents.  It also eliminates the need to maintain and support a second parallel "module-like" system of Conda builds.
